aten/src/ATen/native/vulkan/VulkanAten.cpp

```diff
@@ -0,0 +1,299 @@
+  VTensor& y = vtensor_from_vulkan(output);
+  y.allocateStorage();
+
+#ifdef USE_GLES
+  at::native::vulkan::details::gl::upsample_nearest2d(
+      y, x, ih, iw, oh, ow, in, ic, height_scale, width_scale);
+#else
+  // XXX Not implemented
+  AT_ERROR("upsample_nearest2d_vulkan: Not implemented yet for Vulkan");
+#endif
+  return output;
+}
+
+Tensor vulkan_add(const Tensor& self, const Tensor& other, Scalar alpha) {
+  VTensor& x = vtensor_from_vulkan(self);
+  VTensor& y = vtensor_from_vulkan(other);
+  float a = alpha.to<float>();
+
+  VTensor output = VTensor{self.sizes().vec()};
+  output.allocateStorage();
+
+#ifdef USE_GLES
+  at::native::vulkan::details::gl::add(output, x, y, a);
+#else
+  // XXX Not implemented
+  AT_ERROR("vulkan_add: Not implemented yet for Vulkan");
+#endif
+
+  return new_with_vtensor_vulkan(std::move(output), self.options());
+}
+
+at::Tensor vulkan_convolution(
+    const at::Tensor& input, // Vulkan
+    const at::Tensor& weight, // CPU
+    const at::Tensor& bias, // CPU
+    IntArrayRef padding,
+    IntArrayRef stride,
+    IntArrayRef dilation,
+    int64_t groups) {
+  auto isizes = input.sizes();
+  TORCH_INTERNAL_ASSERT(
+      isizes.size() == 4, "vulkan_convolution: Expected 4-dimensional input");
+  auto wsizes = weight.sizes();
+  TORCH_INTERNAL_ASSERT(
+      wsizes.size() == 4, "vulkan_convolution: Expected 4-dimensional weight");
+  TORCH_INTERNAL_ASSERT(
+      groups == 1,
+      "vulkan_convolution: group convolutions are not supported yet");
+
+  int64_t N = isizes[0];
+  int64_t C = isizes[1];
+  int64_t H = isizes[2];
+  int64_t W = isizes[3];
+
+  int64_t OC = wsizes[0];
+  int64_t KH = wsizes[2];
+  int64_t KW = wsizes[3];
+
+  int64_t PY = padding[0];
+  int64_t PX = padding[1];
+
+  int64_t SY = stride[0];
+  int64_t SX = stride[1];
+
+  int64_t DY = dilation[0];
+  int64_t DX = dilation[1];
+
+  const int KWE = (KW - 1) * DX + 1;
+  const int KHE = (KH - 1) * DY + 1;
+  const int64_t OW = ((W - KWE + 2 * PX) / SX) + 1;
+  const int64_t OH = ((H - KHE + 2 * PY) / SY) + 1;
+
+  auto osizes = std::vector<int64_t>{N, OC, OH, OW};
+
+  const VTensor& vinput = vtensor_from_vulkan(input);
+  VTensor voutput = VTensor{osizes};
+  voutput.allocateStorage();
+
+  float* biasData{};
+  if (bias.defined()) {
+    biasData = bias.template data_ptr<float>();
+  } else {
+    biasData = (float*)std::malloc(sizeof(float) * OC);
+    std::memset(biasData, 0, sizeof(float) * OC);
+  }
+
+  float* weightData = weight.template data_ptr<float>();
+
+#ifdef USE_GLES
+  at::native::vulkan::details::gl::conv2d(
+      voutput,
+      vinput,
+      weightData,
+      KH,
+      KW,
+      biasData,
+      SY,
+      SX,
+      PY,
+      PX,
+      DY,
+      DX,
+      groups);
+#else
+  // XXX Not implemented
+  AT_ERROR("vulkan_convolution: Not implemented yet for Vulkan");
+#endif
+  return new_with_vtensor_vulkan(std::move(voutput), input.options());
+}
+
+} // namespace native
+} // namespace at
+
+#else
```

Comment:
maybe you can change the codegen to avoid ifdefs here? E.g. for CUDA I think we just don't even generate dispatch if compiled without CUDA. We should do the same for vulkan (not I wonder why we didn't do it for mkldnn, lol)